storm-core/test/jvm/org/apache/storm/scheduler/blacklist/TestBlacklistScheduler.java

```diff
@@ -0,0 +1,326 @@
+        hosts.clear();
+        Assert.assertEquals("blacklist", hosts, cluster.getBlacklistedHosts());
+    }
+
+    @Test
+    public void TestReleaseBlacklist() {
+        INimbus iNimbus = new TestUtilsForBlacklistScheduler.INimbusTest();
+
+        Map<String, SupervisorDetails> supMap = TestUtilsForBlacklistScheduler.genSupervisors(3, 4);
+
+        Config config = new Config();
+        config.putAll(Utils.readDefaultConfig());
+        config.put(Config.BLACKLIST_SCHEDULER_TOLERANCE_TIME, 200);
+        config.put(Config.BLACKLIST_SCHEDULER_TOLERANCE_COUNT, 2);
+        config.put(Config.BLACKLIST_SCHEDULER_RESUME_TIME, 300);
+
+        Map<String, TopologyDetails> topoMap = new HashMap<String, TopologyDetails>();
+
+        TopologyDetails topo1 = TestUtilsForBlacklistScheduler.getTopology("topo-1", config, 5, 15, 1, 1, currentTime - 2, true);
+        TopologyDetails topo2 = TestUtilsForBlacklistScheduler.getTopology("topo-2", config, 5, 15, 1, 1, currentTime - 8, true);
+        TopologyDetails topo3 = TestUtilsForBlacklistScheduler.getTopology("topo-3", config, 5, 15, 1, 1, currentTime - 16, true);
+        TopologyDetails topo4 = TestUtilsForBlacklistScheduler.getTopology("topo-4", config, 5, 15, 1, 1, currentTime - 32, true);
+        topoMap.put(topo1.getId(), topo1);
+
+        Cluster cluster = new Cluster(iNimbus, supMap, new HashMap<String, SchedulerAssignmentImpl>(), config);
+        Topologies topologies = new Topologies(topoMap);
+        BlacklistScheduler bs = new BlacklistScheduler(new DefaultScheduler());
+        bs.prepare(config);
+        bs.schedule(topologies, cluster);
+        cluster = new Cluster(iNimbus, TestUtilsForBlacklistScheduler.removeSupervisorFromSupervisors(supMap, "sup-0"), TestUtilsForBlacklistScheduler.assignmentMapToImpl(cluster.getAssignments()), config);
+        bs.schedule(topologies, cluster);
+        cluster = new Cluster(iNimbus, TestUtilsForBlacklistScheduler.removeSupervisorFromSupervisors(supMap, "sup-0"), TestUtilsForBlacklistScheduler.assignmentMapToImpl(cluster.getAssignments()), config);
+        bs.schedule(topologies, cluster);
+        cluster = new Cluster(iNimbus, supMap, TestUtilsForBlacklistScheduler.assignmentMapToImpl(cluster.getAssignments()), config);
+        bs.schedule(topologies, cluster);
+        Set<String> hosts = new HashSet<>();
+        hosts.add("host-0");
+        Assert.assertEquals("blacklist", hosts, cluster.getBlacklistedHosts());
+        topoMap.put(topo2.getId(), topo2);
+        topoMap.put(topo3.getId(), topo3);
+        topoMap.put(topo4.getId(), topo4);
+        topologies = new Topologies(topoMap);
+        cluster = new Cluster(iNimbus, supMap, TestUtilsForBlacklistScheduler.assignmentMapToImpl(cluster.getAssignments()), config);
+        bs.schedule(topologies, cluster);
+        hosts.clear();
+        Assert.assertEquals("blacklist", hosts, cluster.getBlacklistedHosts());
+    }
+
+    @Test
+    public void TestList() {
```

Comment:
This method looks like not easy to read and understand. Could we make this method better? Or at least put more comments for all `if (count == )`.